### ISSUE TITLE
Limit Collision sound length

### ISSUE THREAD

**Issue by Menithal** [highfidelity/hifi#9552]
You can put entire songs into collisions sounds of avatars and objects. Perhaps have a cut off within 1 second?

**Comment by zzmp** [comment]
@Menithal it's just an arbitrary sound. Why do you think it needs a cutoff? What if it's a sustained bell or some other resonating object?

**Comment by Menithal** [comment]
Ok.

Exhibit A. 3x 1 minute 30 secondsof a loop of "Chacarron Macarron" in multiple locations. 
https://www.youtube.com/watch?v=zjk2LPLmHgs

Exhibit B.  1 minute 30 seconds of "We are number one" on an entity and a Entity *
https://www.youtube.com/watch?v=UBm-B5q46qE

Now, knowing the average SL Griefer or even a Gmod one, and access to a throw away account in a hosting solution, such as github

This is /very/ simple to do Doesn't even need much creativity. It doesn't even need one to modify any script, or entity. Multiply this by amount of avatar collisions. Cannot be pinpointed in a crowd using an invisible avatar.  Can be used in a crowded situation, and as soon as it starts playing, it cannot be removed, nor does it go away if you teleport, only goes away if you restart the client.

Sure its limited to 3 active sounds per avatar, but... 

If we remove the no scripts limit: then with a script that sets everyone's collision sounds to such and make  a large expanding collidable sphere that grows by an entity script that everyone edits to grow (to spoof 'last edits') which everyone will bump to and it will just cascade and ruin alot of peoples ears.

Oh and then when combing it with a large entity, and doing a slow velocity, it pitches the time down by the speed, and then plays it back: making a 1 minute collision sound last upto 10-20 minutes.

**Comment by PhilipRosedale** [comment]
this is a good issue to discuss a hard problem.  Just putting a short window on the sound doesn't really solve the problem at all because, for example, you can just write a script that repeatedly bashes two things together.  So I don't think a particular length is worth the trouble it will cause when people then try to exceed it and wonder why it isn't working.

**Comment by Menithal** [comment]
@PhilipRosedale , I do think it is a **limit of time for collisions sounds is good for avatar collisions**, especially as you have a hard 3 collisions sound limit that can be played at the same time for avatars, but it is not a good solution for entity collisions due to the script: But I am more concerned with the avatar collision sounds than with the entity ones.

This is because avatar collisions:
A. Does not Require scripting
B. Does not Require rezing of entity
C. Cannot be removed, even if user is removed from scene.

As of the moment, as you have in earlier meetups stated to me, is that the vulnerabilities related to scripting and entities can be mitigated by having whitelisted scripts and having a trust-based system on the domain. 

But the main issue is that avatar collisions doesnt need either. I really suggest rethinking of this one, if not for collision sounds in general, but atleast avatar collision sounds. I only provided an example of how to make it worse via the scripting example.

**Comment by Menithal** [comment]
Alternatively, Collision sounds should be bound to the object, so that when the source is removed, the collision sound will instantly end. Right now it simply doesnt.

**Comment by BillWindwalker** [comment]
Why not just make the viewer capable of muting unwanted sounds

**Comment by Menithal** [comment]
Because you cannot create a blacklist of sound clips that can be changed upon whim to be of anything hosted on anywhere..
Also: The following
- A system based on whitelist would limit it to the domain.
- You'd need to have a system in for identifying sources of sounds, which currently the system lacks.
- Collision sounds are dislocated from the entity that creates the sound, as it is simply generated on the collision location to play back a sound at a specific pitch

 Thus my suggestion is merely for the shortest route to solution: Have the sound play, but then cut off after X amount of time.

**Comment by BillWindwalker** [comment]
If I remember right there is a mute people list can the item not be muted
by the person list

**Comment by Menithal** [comment]
PersonalMute only works by ignoring session connections by the AudioMixer, which handles the connections between multiple users to decide what you hear and what you cant hear from the server

However, Collision sounds are played back by the client, which when it detects a collision has occurred that the server dictates, plays the AudioInjector which is client side, and doesnt pass through the audio mixer, similar to all audio playback done through the Audio Interface:

This is defined in https://github.com/highfidelity/hifi/blob/5ed376afc90faf18b31856c79ddc84875219f82d/interface/src/avatar/AvatarManager.cpp#L462

There is no definition for when you have the avatar muted. Only when they are ignored, does it work, but by then the sound clip is -already- playing, and only a relog will fix it.

**Comment by BillWindwalker** [comment]
So then it is presume that it is programmed this way and audio length is
not a determining Factor for the sound clips.
It is not a bug nor a flaw so it is a programming feature and if a person
do not wish to hear this feature they can just simply turn down the audo or
turn off the audio.
So why not close off this thread.

**Comment by Menithal** [comment]
Yes **it is a feature, but it is an issue** which I will keep up because It will become a bigger issue as one can abuse it heavily.

If you turn down audio, and turn off audio, you cannot communicate with anyone else. So it is a highly **disruptive** feature, thus it is an issue.

**Comment by BillWindwalker** [comment]
If you are the owner of a Sandbox or world you can have the system auto
return or set up a script to automatically return said item.
Or individuals can move to a private area.
There is a plethora of different scenarios that can move an individual away
from disturbing sound or objects.

**Comment by Menithal** [comment]
@BillWindwalker,
Please read the actual conversation before doing any, assumptions on how to solve it without actually addressing the issue.

Let me say it again:

- You do not need permissions to rez anything, This can be applied directly to the avatar collisions.
- The avatar can be invisible: thus collisions done by the avatar can be done without anyone else knowing what or where the source is.
- The Entity collision that emits the audio can be deleted on collision, yet the audio clip will remain playing
- Even if source is removed, the audio doesn't go away, until everyone restarts their client
- If user teleports during the audio playback, it will follow them where ever they teleport, with the same local positioning. Youd have to teleport to another coordinate for it to be "distant" enough not to be heard.
- If done during an event, it still is highly disruptive, and can leave to the entire venue being abandoned until the audio has stopped playing, or if everyone restarts their client, which is not ideal.

**Comment by Menithal** [comment]
Actually now that I think about it abit more,

@PhilipRosedale there really needs to be an  permission for pushing audio clips to the AudioMixer, inorder to avoid people from just typing Audio playback events into the console and triggering events on the server at other avatar's locations, if permission doesnt exist, simply playback only locally.

**Comment by PhilipRosedale** [comment]
@menithal Sorry I closed too quickly.  Happy to discuss more.  As your last comment added - I think putting a limit on the clip, whether for avatars or objects. will do nothing to address the problem of someone making a lot of noise by repeatedly causing collisions, say with console or by just using a script to knock things together.  

I think the only solution here is to manage some sort of balance which corresponds to the total amount of loudness per unit time (audio 'energy') that a person is allowed to emit into an area.   Would you agree?

**Comment by Menithal** [comment]
@PhilipRosedale  Perhaps, and make avatar collisions abit more costly 
As I mentioned earlier, Currently my main concern I have over Entities, as atleast with entities or scripts you can control it via who can rez entities.

But then the question comes, who owns the entity that creates the collision? what if a script controls the entity passes the control randomly to the rest of the agents in the domain? wouldnt that just drain everyone's energy?

Here is a couple one in order of easiness:

- Setting the Avatar audio collision: As long as the avatar can get to the domain, nothing can stop an avatar playing and spamming a sound clip X or Y times, when entities can be locked and permissions denied from adding new entities. Energy could help in this one as it could be connected with the avatar that collides with something
- Playing a sound clip via console. Energy could help with this one. Can be connected to the avatar as scripts, but could be mitigated with sound injection permissions as well.
- Entity based collision sounds and ones done through distributed scripts. This is where energy is more difficult as everyone would edit to object to collide with objects and cause sound effects, thus draining everyones energy. Can be probably mitigated with the permissions for entity creation
